vehicle/mkvi/software/sensing/sensing.c

```diff
@@ -0,0 +1,73 @@
+/*
+Header:
+    Code for the Sensing Controller Board located in the nostril motherboard and
+the rear sensing box Author:
+    @author Richard Li
+    @date   03/25/2023
+*/
+
+/*----- Includes -----*/
+#include "sensing.h"
+#include "can_api.h"
+#include "log_uart.h"
+#include "utils/timer.h"
+#include <avr/interrupt.h>
+#include <avr/io.h>
+#include <stdio.h>
+#include <string.h>
+#include <util/delay.h>
+
+volatile uint8_t gTimerFlag = 0x01;
+uint8_t gStatusMessage[8] = {};
+volatile uint8_t can_recv_msg[8] = {};
+
+void timer1_isr(void) {
+    send_can = true;
+};
+
+int pcint0_callback(int counter_left) {
+    // Left wheel speed sensor is in pin PB2 -> PORTB
+    sensing.wsl = !gpio_get_pin(WHEEL_SPEED_LEFT);
+    if (sensing.wsl) {
+        counter_left += 1;
+    }
+}
+void pcint2_callback(int counter_right) {
+    // Right wheel speed sensor is in pin PB2 -> PORTB
+    sensing.wsr = !gpio_get_pin(WHEEL_SPEED_RIGHT);
+    if (sensing.wsl) {
+        counter_right += 1;
+    }
+}
+
+int main(void) {
+    can_init_sensing();
+    adc_init();
+    CAN_init(CAN_ENABLED);
+    gpio_set_mode(DEBUG_LED_0, OUTPUT);
+    gpio_set_mode(WHEEL_SPEED_LEFT, INPUT);
+    gpio_set_mode(WHEEL_SPEED_RIGHT, INPUT);
+
+    sei(); // Initializes interrupts
+    gpio_enable_interrupt(WS1);
+    gpio_enable_interrupt(WS2);
+
+    timer_init(&timer1_cfg);
+
+    pcint0_callback();
+    pcint2_callback();
```

Comment:
Little confused by this. Don't these functions need arguments, also why do interrupts need to be called in your main function, shouldn't they just be called based on the gpio channel?

Reply:
Maybe a better structure for this is you had two volatile uint8s that increment on every interrupt call?

Reply:
Not sure about the context here specifically, but I have used this pattern in the past. `pcintN_callback()` are functions that are called by the interrupts for PORTx. The nice thing about having them as standalone functions is that we can also use them as regular functions. Like in the AIR control code, the interrupts are triggered anytime the state of the pins change, but we also call the functions directly to capture the state of the pins during initialization.

Reply:
Yeah, each of these callbacks are triggered whenever the wheel speed sensor sees metal, and the interrupt increments the L + R counters (depending on which pin). I _believe_ this is the correct way to do this?

Reply:
@richardli03 In the interrupts (in these functions) you might need to check to see if the pin has gone high/low before incrementing. We only want to increment each time it goes in _one direction_, but I'm not sure if it's when it goes high or low. Basically we want to increment every time we get to a hole (no metal) and I'm not sure if that's high or low from the WSS. We can test it once we have boards done.